### PR TITLE
Initial support for causal clusters

### PR DESCRIPTION
**Pull request by lutovich**
Available commands:

Initial cluster installation (download package, unpack and configure instances):
```
neoctrl-cluster install --cores 3 --read-replicas 5 3.1.0-M13-beta3 /tmp/cluster
neoctrl-cluster install -c 3 --r 5 3.1.0-M13-beta3 /tmp/cluster
```

Cluster start:
```
neoctrl-cluster start /tmp/cluster
```

Cluster stop:
```
neoctrl-cluster stop /tmp/cluster
```

Re-configuration, start and stop of individual instances could be done using existing commands: `neoctrl-configure`, `neoctrl-start` and `neoctrl-stop`.

Missing:
 * hard kill of instances (will be added to `neoctrl-stop` command)
 * dynamic addition of cluster instances

### CODE REVIEW

**Comment by nigelsmall** [comment]
Could some of those functions be wrapped up as methods of a `Cluster` class?

**Comment by nigelsmall** [comment]
Some additions to the README would also be useful to describe how to use the new tool.

**Comment by nigelsmall** [comment]
Also, is there maybe some value in breaking off some of the config file editing functionality into a generic class or module?

File: boltkit/controller.py [hunk 1]
```diff
@@ -545,3 +600,253 @@ def test():
...
+                raise RuntimeError("Duplicated bolt uri configs found in %s" % config_file_path)
+
+            bolt_uri = _parse_uri("bolt", line)
+
+    return InstanceInfo(http_uri, bolt_uri, path)
+
+
+def _parse_uri(scheme, config_entry):
+    split = config_entry.replace(" ", "").split("=")
+    uri = split[len(split) - 1]
```
Comment (nigelsmall) [inline boltkit/controller.py]: Ditto `partition`.

File: boltkit/controller.py [hunk 2]
```diff
@@ -545,3 +600,253 @@ def test():
...
+        self.path = path
+
+    def http_uri_str(self):
+        return self.http_uri.geturl().strip()
+
+    def bolt_uri_str(self):
+        return self.bolt_uri.geturl().strip()
+
+    def __str__(self):
+        return self.http_uri_str() + " " + self.bolt_uri_str() + " " + self.path
```
Comment (nigelsmall) [inline boltkit/controller.py]: Is this just used for debugging?

File: boltkit/controller.py [hunk 3]
```diff
@@ -287,14 +291,18 @@ def configure(self, properties):
             lines = f_in.readlines()
         with open(config_file_path, "w") as f_out:
             for line in lines:
-                for key, value in properties.items():
+                for key in properties.keys():
                     if line.startswith(key + "=") or \
                             (line.startswith("#") and line[1:].lstrip().startswith(key + "=")):
-                        f_out.write("%s=%s\n" % (key, value))
+                        f_out.write("%s=%s\n" % (key, properties[key]))
+                        del properties[key]
```
Comment (nigelsmall) [inline boltkit/controller.py]: Modifying a collection while iterating through it can be problematic. Why do you need to remove items?

Comment (lutovich) [inline boltkit/controller.py]: It is not needed any more, removed.

File: boltkit/controller.py (hunk 2 above)
Comment (lutovich) [inline boltkit/controller.py]: This is used to format `neoctrl-cluster start` output:

```
http://localhost:7474 bolt://localhost:7687 /tmp/cores/core-0/neo4j-enterprise-3.1.0
http://localhost:7475 bolt://localhost:7688 /tmp/cores/core-1/neo4j-enterprise-3.1.0
http://localhost:7476 bolt://localhost:7689 /tmp/cores/core-2/neo4j-enterprise-3.1.0
http://localhost:7477 bolt://localhost:7690 /tmp/read-replicas/read-replica-0/neo4j-enterprise-3.1.0
http://localhost:7478 bolt://localhost:7691 /tmp/read-replicas/read-replica-1/neo4j-enterprise-3.1.0
```

Then client applications have all info about the cluster and can access it. Does this sound sensible?

**Comment by nigelsmall** [comment]
The organisation or the code is now lots clearer, that's great! Perhaps rename a few of the functions in the `config` module to better describe what they do and/or add some docstring comments.

**Pull request merged**